Add initial Aspire.Hosting.Azure.Network integration

eerhardt opened the pull request:
## Description

This is the first round of supporting Azure virtual networks in Aspire. Currently it supports:

* Creating a virtual network
* Creating subnets
* Creating private endpoints to Storage blobs and queues

Future PRs will enable:

1. More private endpoint support. This is just the core / first resource.
2. Being able to add NAT Gateways and public IPs.

This pull request introduces a new Azure Virtual Network end-to-end sample to the repository, demonstrating how to configure a virtual network, subnets, private endpoints, and secure storage access for containerized applications. It also adds support for Azure Network and Private DNS provisioning packages, and makes minor improvements to existing storage modules.

**New Azure Virtual Network End-to-End Sample:**

* Added the `/playground/AzureVirtualNetworkEndToEnd/` folder, containing `AzureVirtualNetworkEndToEnd.ApiService` and `AzureVirtualNetworkEndToEnd.AppHost` projects to showcase VNet, subnets, private endpoints, and secure storage integration for container apps.
* Implemented infrastructure-as-code modules for container app deployment (`api-containerapp.module.bicep`), managed identity (`api-identity.module.bicep`), and storage role assignments (`api-roles-storage.module.bicep`) to automate resource provisioning and access control. [[1]](diffhunk://#diff-7c8e0581f1039a101f20b1c7a8b58a84b713ca6b68d4f262b7dedd646932f1fbR1-R113) [[2]](diffhunk://#diff-daa7291d3e2f2f738b2a9bd490a1ebfbe140ddbe73e0aa111ac9da78ad18e3e7R1-R17) [[3]](diffhunk://#diff-2d25208a3bd7e5c042f8d5237a8224ad67bdd83a28d94590161673f59808999fR1-R40)
* Added example application code and configuration files for the API service and AppHost, demonstrating usage of Azure Blob and Queue storage within a secured network environment. [[1]](diffhunk://#diff-4ad39c4512aa5983e178770d30c37965555e4e2f1a00da1e0d6293a4ca6a040aR1-R15) [[2]](diffhunk://#diff-d6bd29df42ba6b093ebdb33c3ec54f0a5de5ee5c1b3f1e879cacad5b1431c407R1-R43) [[3]](diffhunk://#diff-da247fd84bbeeb7825d0bad5c251b405b2ff88c50a8086830c7fcc965fd8195eR1-R22) [[4]](diffhunk://#diff-31be5b901c5bb12a6b8d935b11681a93207beb18ed9b86395b4be6509e3a365dR1-R40) [[5]](diffhunk://#diff-d7897fa1e6a8ed47a1861f10ce964018e456812a426d51824e4b6dd1eecf23f5R1-R46) [[6]](diffhunk://#diff-05594a7cdaceb59741c8a617026af535563b43e5bcedb7f1cbf0c627ef877553R1-R9) [[7]](diffhunk://#diff-dd25e9043dc910a8b251d05088b5ec7d232e65ae43909225777ad6301df1cb67R1-R12)

**Support for Azure Network and Private DNS provisioning:**

* Included `Azure.Provisioning.Network` and `Azure.Provisioning.PrivateDns` package references in `Directory.Packages.props` to enable automated VNet and DNS zone management.

## Checklist

- Is this feature complete?
  - [ ] Yes. Ready to ship.
  - [x] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [x] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [x] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [ ] No
- Does the change make any security assumptions or guarantees?
  - [x] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [x] No
  - [ ] No
- Does the change require an update in our Aspire docs?
  - [x] Yes
    - Link to aspire-docs issue (consider using one of the following templates):
      - [New (or update) `doc-idea` template](https://github.com/dotnet/docs-aspire/issues/new?template=02-docs-request.yml)
      - [New `breaking-change` template](https://github.com/dotnet/docs-aspire/issues/new?template=04-breaking-change.yml)
      - [New `diagnostic` template](https://github.com/dotnet/docs-aspire/issues/new?template=06-diagnostic-addition.yml)
  - [ ] No

eerhardt committed:
Add initial Aspire.Hosting.Azure.Network integration

This is the first round of supporting Azure virtual networks in Aspire.

eerhardt committed:
Get some stuff working

eerhardt committed:
Don't add VNet resources in run mode.

Add a delegation annotation so things like ACA can delegate subnets.

tmacam commented on src/Aspire.Hosting.Azure.Network/AzureVirtualNetworkResource.cs:
```diff
@@ -0,0 +1,57 @@
+    public BicepOutputReference NameOutput => new("name", this);
+
+    /// <inheritdoc/>
+    public override ProvisionableResource AddAsExistingResource(AzureResourceInfrastructure infra)
```
For my education: why return `ProvisionableResource` instead of returning `VirtualNetwork` (which is a `ProvisionableResource`)? Is it because that would not work to  to "override" because "C# does not allow covariant return types for normal methods"?

The reason I am asking is that based on my (rather limited) usage of this API, folks end up upper-casting the result of `AddAsExistingResource`. And this feels so unnecessary - and not type-safe  :(

tmacam replied:
Perhaps not the most elegant but something like... 

```suggestion
    public override ProvisionableResource AddAsExistingResource(AzureResourceInfrastructure infra) =>  AddAsExistingResource(AzureResourceInfrastructure infra, out var _);
    
    public ProvisionableResource AddAsExistingResource(AzureResourceInfrastructure infra, out VirtualNetwork existingVNet)
```

eerhardt replied:
>  Is it because that would not work to to "override" because "C# does not allow covariant return types for normal methods"?

Correct. Since this is an override the signature needs to match with the base signature.

eerhardt committed:
Initial support for private endpoints.

eerhardt committed:
Add Azure.Provisioning.PrivateDns and update to latest

eerhardt committed:
Remove unncessary workaround

eerhardt committed:
Temporarily checking in generated bicep

eerhardt committed:
Finish implementing private endpoints.

Gets AzureStorageEndToEnd working with a vnet and private endpoints.

eerhardt committed:
Remove .generated folder and update the generated manifest for the Storage playground app.

eerhardt committed:
Make new Aspire.Hosting.Azure APIs experimental

eerhardt committed:
Add unit tests.

Remove NatGateway (for now). This will come in a separate PR.

Clean up some code.

eerhardt committed:
More clean up

eerhardt commented on src/Aspire.Hosting.Azure.Network/README.md:
```diff
@@ -0,0 +1,105 @@
+
+### Complete example with outbound connectivity
+
+This example creates a Virtual Network with a subnet that has outbound internet connectivity via a NAT Gateway:
```
I removed the NAT gateway stuff for now. that will be a follow up PR.

eerhardt committed:
Move to a new playground app specific to VNets

eerhardt committed:
Fix AddSubnet API

eerhardt committed:
Rename AddAzurePrivateEndpoint and hang it off of AzureSubnetResource.

eerhardt committed:
Don't add PrivateEndpointTargetAnnotation in run mode. Also walk the hierarchy until you reach the root resource.

eerhardt committed:
Add WithSubnet API for ACA environments and improve delegation naming

- Add IAzureDelegatedSubnetResource interface and DelegatedSubnetAnnotation to Aspire.Hosting.Azure
- Implement IAzureDelegatedSubnetResource on AzureContainerAppEnvironmentResource
- Add WithSubnet<T> extension method in Network package for configuring subnet delegation
- Update ACA Bicep generation to read DelegatedSubnetAnnotation and configure VnetConfiguration
- Use serviceName for delegation name in Bicep output (e.g., 'Microsoft.App/environments')
- Mark new public APIs with [Experimental("ASPIREAZURE003")]
- Add unit test verifying ACA and VNet Bicep output with subnet delegation

eerhardt committed:
Separate and reuse Private DNS Zones across private endpoints

- Add AzurePrivateDnsZoneResource for shared DNS Zone management
- Add AzurePrivateDnsZoneVNetLinkResource for VNet-to-zone linking
- DNS Zones are cached at builder level and reused per zone name
- VNet Links tracked on DNS Zone to avoid duplicates
- PE Bicep now references existing DNS Zone instead of creating inline
- Add tests for DNS Zone reuse scenarios
- Add InternalsVisibleTo for test access to internal types

karolz-ms commented on tests/Aspire.Hosting.Azure.Tests/AzureVirtualNetworkExtensionsTests.cs:
```diff
@@ -0,0 +1,107 @@
+        // In run mode, the subnet should not be added to the builder's resources
+        Assert.DoesNotContain(subnet.Resource, builder.Resources);
+    }
+}
```
Maybe I missed it, but I do not see a test that calls `WithSubnet()` extension method. Should we add one?

karolz-ms reviewed (commented):
Looks great overall, just a few questions

eerhardt commented on src/Aspire.Hosting.Azure/IAzureDelegatedSubnetResource.cs:
```diff
@@ -0,0 +1,19 @@
+namespace Aspire.Hosting.Azure;
+
+/// <summary>
+/// Represents an Azure resource that requires a subnet with service delegation.
```
Fixed the wording to "supports".

eerhardt committed:
Address PR feedback

mitchdenny commented:
/deployment-tests

mitchdenny reviewed (approved):
Looks good! Nice foundation for Azure networking support.

A few suggestions for follow-up work (already noted in comments):
- Consider adding a `ParameterResource` overload for `AddSubnet` to allow per-environment address prefixes
- E2E deployment tests would be valuable once there's a way to validate private network connectivity (e.g., via ACI in the VNet)
- Future consideration: Application Gateway / Azure Front Door support for the "private backend + public ingress" pattern
